README.md
README
======

[![Build Status](https://travis-ci.org/blablacar/worker-lib.png)](https://travis-ci.org/blablacar/worker-lib)


Description
-----------

Simple library to consume your AMQP messages.

Installation
------------

Add the library to your application's `composer.json` file:

```json
{
    "require": {
        "blablacar/worker-lib": "1.0.*"
    }
}
```

Launch tests
------------

Once you have a working RabbitMQ running on local and PHPUnit installed, you
can simply launch

    ./tests.sh

Tests will delete you "/" vhost.


CHANGELOG.md
CHANGELOG
=========

DONE
----

* [2013-12-02] Avoid lost message when publishing to an unexisting exchange
  then to a valid exchange (used transactions)
* Replace "consume" by "get" to avoid closed connection (with configurable
  poll-interval)

TODO
----

* Handle PHP Catchable Fatal errors ?
* Add profiling
* Refactor SignalHandler
* Clean Manager::consume & Wrapper (use encapsulation)


tests.sh
#!/bin/bash

./tests/bin/init_rabbit.sh

phpunit


tests/bootstrap.php
<?php

function includeIfExists($file)
{
    if (file_exists($file)) {
        return include $file;
    }
}

if ((!$loader = includeIfExists(__DIR__.'/../vendor/autoload.php')) && (!$loader = includeIfExists(__DIR__.'/../../../autoload.php'))) {
    echo 'You must set up the project dependencies, run the following commands:'.PHP_EOL.
        'curl -s http://getcomposer.org/installer | php'.PHP_EOL.
        'php composer.phar install'.PHP_EOL;
    exit(1);
}

$loader->add('Blablacar\Worker\\', __DIR__);


tests/Blablacar/Worker/AMQP/ManagerTest.php
<?php

namespace Blablacar\Worker\AMQP;

use Blablacar\Worker\AMQP\Manager;
use Blablacar\Worker\AMQP\Consumer\Context;
use Blablacar\Worker\AMQP\Consumer\Wrapper;

class ManagerTest extends \PHPUnit_Framework_TestCase
{
    public function testInstanceOf()
    {
        $manager = new Manager(new \AMQPConnection());
        $this->assertInstanceOf('Blablacar\Worker\AMQP\Manager', $manager);
    }

    public function testGetConfig()
    {
        $manager = new Manager(new \AMQPConnection());
        $this->assertEquals('guest[:guest]@localhost/:5672', $manager->getConfig());
    }

    public function testConsumeWithConsumerInterface()
    {
        $manager = new Manager(new \AMQPConnection());

        $consumer = $this->getMock('Blablacar\Worker\AMQP\Consumer\ConsumerInterface');
        $consumer
            ->expects($this->exactly(3))
            ->method('__invoke')
            ->with(
                $this->anything(),
                $this->anything(),
                $this->anything()
            )
            ->will($this->onConsecutiveCalls(true, true, false))
        ;

        $manager->consume('blablacar_worker_queue_test', $consumer, null, AMQP_NOPARAM);
    }

    public function testConsumeWithWrapper()
    {
        $manager = new Manager(new \AMQPConnection());

        $consumer = $this->getMock('Blablacar\Worker\AMQP\Consumer\ConsumerInterface');
        $consumer
            ->expects($this->exactly(3))
            ->method('__invoke')
            ->with(
                $this->anything(),
                $this->anything(),
                $this->anything()
            )
            ->will($this->returnValue(true))
        ;

        $manager->consume('blablacar_worker_queue_test', new Wrapper($consumer), new Context(300, 3));
    }

    public function testDeleteQueue()
    {
        $manager = new Manager(new \AMQPConnection());
        $isDeleted = $manager->deleteQueue('blablacar_worker_delete_queue_test');

        $this->assertTrue($isDeleted);
    }

    public function testPublish()
    {
        $manager = new Manager(new \AMQPConnection());
        $isPublished = $manager->publish('blablacar_worker_exchange_test', 'my message', 'test');

        $this->assertTrue($isPublished);
    }

    public function testPublishMessageWithUnknownExchangeThenWithKnownExchangeTheConsumeIt()
    {
        $this->markTestSkipped('This test don\'t work due to php-amqp lib');

        $manager = new Manager(new \AMQPConnection());
        $isPublished = $manager->publish('unknown_exchange', 'my message', 'test');
        $this->assertFalse($isPublished);
        $isPublished = $manager->publish('blablacar_worker_empty_exchange_test', 'my message', 'test');
        $this->assertTrue($isPublished);

        $consumer = $this->getMock('Blablacar\Worker\AMQP\Consumer\ConsumerInterface');
        $consumer
            ->expects($this->once())
            ->method('__invoke')
            ->with(
                $this->anything(),
                $this->anything(),
                $this->anything()
            )
            ->will($this->returnValue(false))
        ;

        $manager->consume('blablacar_worker_empty_queue_test', new Wrapper($consumer), new Context(4, 1));
    }
}


src/Blablacar/Worker/AMQP/Manager.php
<?php

namespace Blablacar\Worker\AMQP;

use Blablacar\Worker\AMQP\Consumer\ConsumerWrapper;
use Blablacar\Worker\AMQP\Consumer\ConsumerInterface;
use Blablacar\Worker\AMQP\Consumer\Context;

class Manager
{
    protected $connection;
    protected $channel;

    protected $startTime;

    public function __construct(\AMQPConnection $connection)
    {
        $this->connection = $connection;
    }

    /**
     * getConfig
     *
     * @return string
     */
    public function getConfig()
    {
        return sprintf(
            '%s[:%s]@%s%s:%d',
            $this->connection->getLogin(),
            $this->connection->getPassword(),
            $this->connection->getHost(),
            $this->connection->getVhost(),
            $this->connection->getPort()
        );
    }

    /**
     * publish
     *
     * If the AMQP_MANDATORY flag is not passed, and the publication failed, an
     * exception is thrown. Otherwise: it just return false
     *
     * @see http://us3.php.net/manual/en/amqp.constants.php
     *
     * @param string $exchange
     * @param string $message
     * @param string $routingKey
     * @param int    $flags
     * @param array  $attributes
     *
     * @throw \AMQPChannelException
     *
     * @return boolean
     */
    public function publish($exchange, $message, $routingKey, $flags = AMQP_NOPARAM, array $attributes = array())
    {
        $exchange = $this->getExchange($exchange);

        return $exchange->publish($message, $routingKey, $flags, $attributes);
    }

    /**
     * consume
     *
     * @param string  $queue
     * @param mixed   $consumer
     * @param Context $context
     * @param int     $flags
     *
     * @return void
     */
    public function consume($queue, $consumer, Context $context = null, $flags = AMQP_NOPARAM)
    {
        if (null === $context) {
            $context = new Context();
        }

        if ($consumer instanceof ConsumerInterface) {
            $consumer->preProcess($context);
        }

        $this->startTime = time();

        $queue = $this->getQueue($queue);

        $continue = true;
        while ($continue) {
            $envelope = $queue->get($flags);

            if (!$envelope) {
                usleep($context->getPollInterval());
            } else {
                $continue = $consumer($envelope, $queue, $context);
            }

            $elapsedTime = microtime(true)-$this->startTime;
            if ($elapsedTime >= $context->getMaxExecutionTime()) {
                $context->output(sprintf(
                    '<info>Maximum time exceeded. Exiting after <comment>%.2fs</comment>.</info>',
                    $elapsedTime
                ));

                return false;
            }
        }

        if ($consumer instanceof ConsumerInterface) {
            $consumer->postProcess($context);
        }
    }

    /**
     * createExchange
     *
     * @param string $name
     *
     * @return void
     */
    public function createExchange($name, $type, $flags)
    {
        $ex = $this->getExchange($name);
        $ex->setType(AMQP_EX_TYPE_DIRECT);
        $ex->setFlags($flags);
        $ex->declareExchange();

        return $ex;
    }

    /**
     * createQueue
     *
     * @param string $name
     *
     * @return void
     */
    public function createQueue($name, $flags = AMQP_NOPARAM, array $arguments = array())
    {
        $queue = $this->getQueue($name);
        $queue->setFlags($flags);
        $queue->setArguments($arguments);
        $queue->declareQueue();

        return $queue;
    }

    /**
     * deleteQueue
     *
     * @param strin $name
     *
     * @return boolean
     */
    public function deleteQueue($name)
    {
        if (strlen($name) == 0) {
            return false;
        }

        try {
            $queue = $this->getQueue($name);

            return $queue->delete(AMQP_IFEMPTY|AMQP_IFUNUSED);
        } catch (\AMQPException $e) {
            return false;
        }
    }

    /**
     * getExchange
     *
     * @param string $name
     *
     * @return \AMQPExchange
     */
    protected function getExchange($name)
    {
        $channel = $this->connect();

        $exchange = new \AMQPExchange($channel);
        $exchange->setName($name);

        return $exchange;
    }

    /**
     * getQueue
     *
     * @param string $name
     *
     * @return \AMQPQueue
     */
    protected function getQueue($name)
    {
        $channel = $this->connect();

        $queue = new \AMQPQueue($channel);
        $queue->setName($name);

        return $queue;
    }

    /**
     * connect
     *
     * @return void
     */
    protected function connect()
    {
        if (!$this->connection->isConnected()) {
            $this->connection->connect();
            $this->channel = null;
        }

        if (null === $this->channel) {
            $this->channel = new \AMQPChannel($this->connection);
        }

        return $this->channel;
    }
}


src/Blablacar/Worker/Util/SignalHandler.php
<?php

namespace Blablacar\Worker\Util;

class SignalHandler
{
    static protected $handledSignals      = array(SIGTERM, SIGINT, SIGQUIT);
    static protected $shouldExitForSignal = false;

    public static function start()
    {
        if (!extension_loaded('pcntl')) {
            return;
        }

        foreach (self::$handledSignals as $signal) {
            pcntl_signal($signal, function ($signal) {
                SignalHandler::$shouldExitForSignal = true;
            });
        }
    }

    public static function haveToStop()
    {
        if (!extension_loaded('pcntl')) {
            return false;
        }

        pcntl_signal_dispatch();

        foreach (self::$handledSignals as $signal) {
            pcntl_signal($signal, SIG_DFL);
        }

        return self::$shouldExitForSignal;
    }
}


src/Blablacar/Worker/AMQP/Consumer/Context.php
<?php

namespace Blablacar\Worker\AMQP\Consumer;

use Symfony\Component\Console\Output\OutputInterface;

class Context
{
    protected $maxExecutionTime = 300;
    protected $maxMessages      = 300;
    protected $useSigHandler    = false;
    protected $requeueOnError   = false;
    protected $pollInterval     = 500000; // microseconds

    protected $output;

    public function __construct(
        $maxExecutionTime = 300,
        $maxMessages      = 300,
        $useSigHandler    = false,
        $requeueOnError   = false,
        $pollInterval     = 500000
    )
    {
        $this->maxExecutionTime = $maxExecutionTime;
        $this->maxMessages      = $maxMessages;
        $this->useSigHandler    = $useSigHandler;
        $this->requeueOnError   = $requeueOnError;
        $this->pollInterval     = $pollInterval;
    }

    /**
     * output
     *
     * @param string $output
     *
     * @return void
     */
    public function output($output)
    {
        if (null === $this->output) {
            return;
        }

        $this->output->writeln($output);
    }

    /**
     * debug
     *
     * @param string $output
     *
     * @return void
     */
    public function debug($output)
    {
        if (null === $this->output) {
            return;
        }

        if ($this->output->getVerbosity() < OutputInterface::VERBOSITY_DEBUG) {
            return;
        }

        $this->output->writeln($output);
    }

    public function getMaxExecutionTime()
    {
        return $this->maxExecutionTime;
    }

    public function setMaxExecutionTime($maxExecutionTime)
    {
        $this->maxExecutionTime = $maxExecutionTime;
    }

    public function getMaxMessages()
    {
        return $this->maxMessages;
    }

    public function setMaxMessages($maxMessages)
    {
        $this->maxMessages = $maxMessages;
    }

    public function getUseSigHandler()
    {
        return $this->useSigHandler;
    }

    public function setUseSigHandler($useSigHandler)
    {
        $this->useSigHandler = $useSigHandler;
    }

    public function getRequeueOnError()
    {
        return $this->requeueOnError;
    }

    public function setRequeueOnError($requeueOnError)
    {
        $this->requeueOnError = $requeueOnError;
    }

    public function getOutput()
    {
        return $this->output;
    }

    public function setOutput(OutputInterface $output)
    {
        $this->output = $output;
    }

    public function getPollInterval()
    {
        return $this->pollInterval;
    }

    public function setPollInterval($pollInterval)
    {
        $this->pollInterval = $pollInterval;
    }
}


src/Blablacar/Worker/AMQP/Consumer/Wrapper.php
<?php
declare(ticks = 1);

namespace Blablacar\Worker\AMQP\Consumer;

use Blablacar\Worker\Util\SignalHandler;
use Psr\Log\LoggerInterface;

/**
 * Wrapper
 *
 * @TODO: Refactor the __invoke method !
 */
class Wrapper implements ConsumerInterface
{
    protected $consumer;
    protected $logger;

    protected $nbMessagesProcessed = 0;

    public function __construct($consumer, LoggerInterface $logger = null)
    {
        $this->consumer = $consumer;
        $this->logger   = $logger;
    }

    /**
     * {@inheritDoc}
     */
    public function preProcess(Context $context = null)
    {
        if (null === $context) {
            return;
        }

        $context->output(sprintf(
            '<info>Run worker (pid: <comment>%d</comment>. Consume <comment>%d messages</comment> or stop after <comment>%ds</comment>.</info>',
            getmypid(),
            $context->getMaxMessages(),
            $context->getMaxExecutionTime()
        ));
    }

    /**
     * {@inheritDoc}
     */
    public function postProcess(Context $context = null)
    {
    }

    /**
     * {@inheritDoc}
     */
    public function __invoke(\AMQPEnvelope $envelope, \AMQPQueue $queue, Context $context = null)
    {
        $currentStartTime = microtime(true);

        if ($context->getUseSigHandler()) {
            SignalHandler::start();
        }

        try {
            $consumer = $this->consumer;
            $consumer($envelope, $queue, $context);

            $queue->ack($envelope->getDeliveryTag());

            $context->output(sprintf(
                '<comment>ACK [%s]. Duration <info>%.2fs</info>. Memory usage: <info>%.2f Mo</info></comment>',
                $envelope->getDeliveryTag(),
                microtime(true)-$currentStartTime,
                round(memory_get_usage()/1024/1024, 2)
            ));
        } catch (\Exception $e) {
            $queue->nack($envelope->getDeliveryTag(), $context->getRequeueOnError()? AMQP_REQUEUE : null);
            $context->output(sprintf(
                '<error>NACK [%s].</error>',
                $envelope->getDeliveryTag()
            ));

            if (null !== $this->logger) {
                $this->logger->error(sprintf(
                    'Error occured with queue "%s". Message: %s. Exception: %s',
                    $queue->getName(),
                    $e->getMessage(),
                    $e->getTraceAsString()
                ));
            }

            if (null !== $context->getOutput() && $context->getOutput()->getVerbosity() >= 2) {
                throw $e;
            }
        }

        if (++$this->nbMessagesProcessed >= $context->getMaxMessages()) {
            $context->output(sprintf(
                '<info>Max messages reached. Exiting after processing <comment>%d messages</comment>.</info>',
                $this->nbMessagesProcessed
            ));

            return false;
        }

        if ($context->getUseSigHandler() && SignalHandler::haveToStop()) {
            $context->output(sprintf(
                '<info>Signal received. Exiting after processing <comment>%d messages</comment>.</info>',
                $this->nbMessagesProcessed
            ));

            return false;
        }

        return true;
    }
}


src/Blablacar/Worker/Command/WorkerCommand.php
<?php

namespace Blablacar\Worker\Command;

use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Blablacar\Worker\AMQP\Consumer\Context;

abstract class WorkerCommand extends Command
{
    /**
     * getManager
     *
     * @param string $connection
     *
     * @return Manager
     */
    abstract protected function getManager($connection = 'default');

    /**
     * getConsumer
     *
     * @return ConsumerInterface
     */
    abstract protected function getConsumer(InputInterface $input, OutputInterface $output);

    /**
     * {@inheritDoc}
     */
    protected function configure()
    {
        $this
            ->setName('consume')
            ->setDescription('Consume a queue')
            ->addArgument('queue', InputArgument::REQUIRED, 'Queue to consume')
            ->addArgument('connection', InputArgument::OPTIONAL, 'Connection to use', 'default')
            ->addOption('timeout', null, InputOption::VALUE_REQUIRED, 'Timeout (seconds) before exit', 300)
            ->addOption('max-messages', null, InputOption::VALUE_REQUIRED, 'Max messages to process before exit', 300)
            ->addOption('no-sighandler', null, InputOption::VALUE_NONE, 'Disable signal handlers')
            ->addOption('requeue-on-error', null, InputOption::VALUE_NONE, 'Requeue in the same queue on error')
            ->addOption('poll-interval', null, InputOption::VALUE_REQUIRED, 'Poll interval (in micro-seconds)', 500000)
        ;
    }

    /**
     * {@inheritDoc}
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $queue = $input->getArgument('queue');
        $connection = $input->getArgument('connection');

        // Manager
        try {
            $manager = $this->getManager($connection);
            if (OutputInterface::VERBOSITY_DEBUG === $output->getVerbosity()) {
                $output->writeln(sprintf(
                    'Use connection: "%s"',
                    $manager->getConfig()
                ));
            }
        } catch (\Exception $e) {
            $output->writeln(sprintf(
                '<error>No manager "%s".</error>',
                $connection
            ));

            return 1;
        }

        // Consumer
        $consumer = $this->getConsumer($input, $output);

        // Context
        $context = $this->getContext($input, $output);

        $manager->consume($queue, $consumer, $context);
    }

    /**
     * getContext
     *
     * @param InputInterface $input
     * @param OutputInterface $output
     *
     * @return Context
     */
    protected function getContext(InputInterface $input, OutputInterface $output)
    {
        $context = new Context(
            $input->getOption('timeout'),
            $input->getOption('max-messages'),
            !$input->getOption('no-sighandler'),
            $input->getOption('requeue-on-error'),
            $input->getOption('poll-interval')
        );
        $context->setOutput($output);

        return $context;
    }
}
